### PR TITLE
Use Vercel action for continuous documentation

### PR DESCRIPTION
**Pull request by seisman**
**Description of proposed changes**

For Continous Documentation (i.e., previewing documentation in PRs), we're currently using the [Vercel App](https://github.com/apps/vercel). To make the app work, we need two configuration files in the repository root directory, `package.json` and `vercel.json`. The Vercel App works well for PyGMT, but is more challenging for GMT (https://github.com/GenericMappingTools/gmt/issues/4848#issuecomment-785391790).

This PR adds a new workflow using the [vercel-action](https://github.com/marketplace/actions/vercel-action) action.

The new Vercel action has the following pros:

1. `package.json` and `vercel.json` are no longer needed. Two fewer files in the root directory
2. The workflow is easier to understand and maintain. We just need to build the documentation as usual and run the vercel action to deploy it to vercel.
3. The old vercel app shows "View deployment" button for each build. It adds too much vertical spaces, making a PR page much longer. The new vercel action doesn't have such a problem.
image

The new Vercel action also has some cons:

1. After the first build finishes, the action creates a comment, showing the preview URL. For the following builds, the action updates the preview URL in the comment. So PR authors and maintainers will be notified once by the action. It's noisier than the old Vercel App.
image

2. The comment only shows the preview URL of the latest build, so it's not easy to compare the documentation of two builds. It's still possible to get old preview URLs by checking the workflow output.

3. The workflow takes ~7 minutes, slightly slower than the old Vercel App (~6 minutes).

**Notes:**

You may notice that the Vercel App also creates a comment and still show the "View deployment" buttons, even though we already remove the `package.json` and `vercel.json` files. That's simply because we still have the Vercel App installed and enabled for this repository. We should diable the Vercel App if we decide to use the new workflow.

image
image

**TODO**

- [ ] Update CONTRIBUTING.md and MAINTENANCE.md

Fixes #

**Reminders**

- [ ] Run `make format` and `make check` to make sure the code follows the style guide.
- [ ] Add tests for new features or tests that would have caught the bug that you're fixing.
- [ ] Add new public functions/methods/classes to `doc/api/index.rst`.
- [ ] Write detailed docstrings for all functions/methods.
- [ ] If adding new functionality, add an example to docstrings or tutorials.

**Slash Commands**

You can write slash commands (`/command`) in the first line of a comment to perform
specific operations. Supported slash commands are:

- `/format`: automatically format and lint the code
- `/test-gmt-dev`: run full tests on the latest GMT development version

### CODE REVIEW

**Comment by seisman** [comment]
Ping @GenericMappingTools/python and @GenericMappingTools/python-maintainers for comments about the new workflow, especially that if you feel that the comment below is annoying.

image

**Comment by meghanrjones** [comment]
I think the new comment structure is a worthwhile trade off for fewer files and easier maintenance. It will be nice to have consistency with continuous documentation between PyGMT and GMT.

**Review by weiji14 — COMMENTED** [review]
A few things to address documentation wise on the  ["Continuous Documentation" section in MAINTENANCE.md](https://github.com/GenericMappingTools/pygmt/blob/v0.3.0/MAINTENANCE.md#continuous-documentation):

1. Remove references to `package.json` and `vercel.json`.
2. Mention how the secrets `VERCEL_ORG_ID`, `VERCEL_PROJECT_ID` and `VERCEL_TOKEN` are created, and how they can be updated (i.e. link to the Vercel page).

Optional points:

- Can we set up a matrix build on all 3 OSes in this action? I.e. remove the "Build the documentation" section in `ci_tests.yaml` and put it here instead.

Rationale: Saves ~4min of CI time. I wanted to separate the tests from the documentation builds in #830, but maybe we can do it in this "Continuous Documentation" workflow instead. I know Leo mentioned using the documentation build as a 'test' before (lazy to dig up that comment), but our >20 min of CI tests is getting wayy too long.

File: .github/workflows/continuous-documentation.yml [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+      - name: Deploy documentation to Vercel
+        uses: amondnet/vercel-action@v20.0.0
+        with:
+          # https://vercel.com/account/tokens
+          vercel-token: ${{ secrets.VERCEL_TOKEN }}
+          # vercel-org-id and vercel-project-id are generated using vercel CLI
+          vercel-org-id: ${{ secrets.VERCEL_ORG_ID }}
+          vercel-project-id: ${{ secrets.VERCEL_PROJECT_ID }}
+          github-token: ${{ secrets.GITHUB_TOKEN }}
+          github-comment: true
```
[the remark this answers is not in the document]
Comment (seisman) [reply .github/workflows/continuous-documentation.yml:92]: > I prefer not having a github comment every push,

No, there is only one comment. Following pushes/build will update the existing comment, not create a new one. You can check the history of this comment https://github.com/GenericMappingTools/pygmt/pull/964#issuecomment-785520078

[the remark this answers is not in the document]
Comment (weiji14) [reply .github/workflows/continuous-documentation.yml:92]: Ah yes. Then it's fine. Only looked at your screenshot in https://github.com/GenericMappingTools/pygmt/pull/964#issuecomment-785644413 (which must have been an old config?) and I thought it was a unique URL for each push.

**Comment by seisman** [comment]
> A few things to address documentation wise on the ["Continuous Documentation" section in MAINTENANCE.md](https://github.com/GenericMappingTools/pygmt/blob/v0.3.0/MAINTENANCE.md#continuous-documentation):
> 
> 1. Remove references to `package.json` and `vercel.json`.

Will do.

> 2. Mention how the secrets `VERCEL_ORG_ID`, `VERCEL_PROJECT_ID` and `VERCEL_TOKEN` are created, and how they can be updated (i.e. link to the Vercel page).

What about writing the instructions in the workflow file, so the workflow is self-contained and the maintenance guide is short.

**Comment by weiji14** [comment]
> > 1. Mention how the secrets `VERCEL_ORG_ID`, `VERCEL_PROJECT_ID` and `VERCEL_TOKEN` are created, and how they can be updated (i.e. link to the Vercel page).
> 
> What about writing the instructions in the workflow file, so the workflow is self-contained and the maintenance guide is short.

Yep, an inline comment with the Vercel documentation link works too!

File: .github/workflows/continuous-documentation.yml [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+          activate-environment: pygmt
+          python-version: 3.9
+          channels: conda-forge
+          miniconda-version: "latest"
+
+      - name: Install dependencies
+        run: |
+          conda install gmt=6.1.1 numpy pandas xarray netcdf4 packaging \
+                        ipython myst-parser sphinx sphinx-copybutton \
+                        sphinx-gallery sphinx_rtd_theme=0.4.3
```
Comment (seisman) [inline .github/workflows/continuous-documentation.yml:54]: Here I only install the packages required for documentation, to save the building time.

**Comment by seisman** [comment]
> Optional points:
> 
> * Can we set up a matrix build on all 3 OSes in this action? I.e. remove the "Build the documentation" section in `ci_tests.yaml` and put it here instead.
> 
> Rationale: Saves ~4min of CI time. I wanted to separate the tests from the documentation builds in #830, but maybe we can do it in this "Continuous Documentation" workflow instead. I know Leo mentioned using the documentation build as a 'test' before (lazy to dig up that comment), but our >20 min of CI tests is getting wayy too long.

I'm OK with separate workflows for docs and tests. We can try it in a separate PR.

File: MAINTENANCE.md [hunk 3]
```diff
@@ -84,16 +84,19 @@ This workflow is ran to publish wheels to PyPI and TestPyPI (for testing only).
...
-change the default behaviour at https://vercel.com/docs/configuration.
-The actual script `package.json` is used by Vercel to install the necessary packages,
-build the documentation, copy the files to a 'public' folder and deploy that to the web,
-see https://vercel.com/docs/build-step.
+We use the [Vercel](https://vercel.com/) service and the
+[vercel-action](https://github.com/marketplace/actions/vercel-action) action to
+preview changes made to our documentation website every time we make a commit
+in a pull request. The workflow `continuous-documentation.yml` builds and
+deploys the documentation to Vercel and create a comment with the latest URL
+of deployment preview.
```
Comment (weiji14) [inline MAINTENANCE.md:99]: The bot's comment seems to only point to a single URL (e.g. https://pygmt-git-vercel-action-gmt.vercel.app in this PR) and now a unique one.

```suggestion
deploys the documentation to Vercel. The vercel bot will automatically make a
comment with a URL to preview the deployed documentation for that pull request.
```

File: .github/workflows/continuous-documentation.yml [hunk 4]
```diff
@@ -83,9 +83,15 @@ jobs:
...
         with:
-          # https://vercel.com/account/tokens
+          # Generate a vercel token at https://vercel.com/account/tokens,
+          # and save the value in the secrets setting in the repository.
           vercel-token: ${{ secrets.VERCEL_TOKEN }}
-          # vercel-org-id and vercel-project-id are generated using vercel CLI
+          # https://github.com/marketplace/actions/vercel-action#project-linking
+          # Install the Vercel CLI and run `vercel` in the project directory.
+          # After linking the repository with a Vercel project, `vercel` will
+          # create a `.vercel` directory containing both the organization (vercel-org-id)
```
Comment (weiji14) [inline .github/workflows/continuous-documentation.yml:92]: I didn't checkout this branch locally, but is the a `.vercel` directory we need to add to `.gitignore`?

Comment (seisman) [reply .github/workflows/continuous-documentation.yml:92]: Running the `vercel` command will create the `.vercel` directory, with a `project.json` file inside. The `project.json` file contains the id information:
```
{"orgId":"example_org_id","projectId":"example_project_id"}
```

The directory and the file are required if I want to deploy the documentation (maybe using a command like `vercel deploy`) from my own computer. But for GitHub Actions, we only need to provide the ids and the action should run the command for us. So, `.vercel` are not added to the repository, and I also deleted the `.vercel` directory after adding the IDs to secrets.

Comment (weiji14) [reply .github/workflows/continuous-documentation.yml:92]: Ok, no need to add to .gitignore then.

**Review by weiji14 — APPROVED** [review]
14 lines to 99 lines of code to get rid of 2 files. Hope this is worth it :laughing:

**Comment by seisman** [comment]
> P.S. Remember to uninstall the Vercel App later.

Just disabled the App in this repository, because it's still used by https://github.com/GenericMappingTools/website.

**Pull request merged**